### DISCUSSION TITLE
Client IP Address — Edge Functions

### DISCUSSION THREAD

**Discussion by WilliamScott-AdventGX** [supabase/supabase#7884] [Questions] [answered]
As far as I can tell, an edge function can not obtain the IP address of the client who invoked it.

Headers `X-Forwarded-From`, `host`, and deno's STD serve function's parameter `connInfo` all return supabase related IP addresses or the local machine address of the edge function.

Is anyone aware if this is possible or impossible?
We are needing to keep a log of IP addresses that invoked the function.

Thanks

**Comment by nounder** [comment]
This is critical for many location-based app, such as commerce or content providers. Without access to client IP it is impossible to do any location-based features for the user, such as currency/country detection.

**Comment by nounder** [comment]
As a workaround, If you are using Next.js, you can implement your own Next.js edge API endpoint as following:

```
import type { NextRequest } from "next/server";

export const config = {
  runtime: "experimental-edge",
};

export default async function handler(req: NextRequest) {
  let ip = req.ip;

  // you can use req.geo on some hosted platform, like Vercel (only for paid plans)

  // IP not available probably because we're on our workstation.
  // Get its external IP to simulate real world.
  if (!ip) {
    const { origin } = await fetch("https://httpbin.org/ip").then((v) =>
      v.json()
    );

    ip = origin;
  }

  const ipInfo = await fetchIpInfo(ip);

  return json(ipInfo);
}

async function fetchIpInfo(ip) {
  // Use external API, like: ipregistry.co
}

function json(data) {
  return new Response(
    JSON.stringify({
      ...data,
    }),
    {
      status: 200,
      headers: {
        "content-type": "application/json",
      },
    }
  );
}
```

**Comment by laktek** [comment, marked as answer]
@WilliamScott-AdventGX @r-ae-i Thanks for bringing this up! We've made an update that should populate the ` X-Forwarded-For` header with the request's client IP correctly.

Here's an example snippet on how you can detect user's location based on IP in an edge function:

```
import { serve } from "https://deno.land/std@0.131.0/http/server.ts"

function ips(req: Request) {
  return req.headers.get("x-forwarded-for")?.split(/\s*,\s*/);
}

serve(async (req) => {
  const clientIps = ips(req) || [''];
  const res = await fetch(`https://ipinfo.io/${clientIps[0]}?token=XXXX`, {
      headers: { 'Content-Type': 'application/json'}});
  const { city, country } = await res.json();

  return new Response(
    JSON.stringify(`You're accessing from ${city}, ${country}`),
    { headers: { "Content-Type": "application/json" } },
  )
})
```

**Reply by idevelop** [reply]
@laktek I think the header is `X-Forwarded-For` - it's written correctly in your code snippet but not in your first paragraph.

**Reply by laktek** [reply]
@idevelop Thanks! updated the comment.

**Reply by torkleyy** [reply]
I'm using exactly this code, but it seems the `x-forwarded-for` header isn't always set - half of the time it's just empty.
